spinalcordtoolbox/scripts/sct_label_vertebrae.py

```diff
@@ -31,39 +31,33 @@
 from spinalcordtoolbox.scripts import sct_straighten_spinalcord, sct_apply_transfo
 
 
-# PARAMETERS
-class Param:
-    # The constructor
-    def __init__(self):
-        self.shift_AP = 32  # shift the centerline towards the spine (in voxel).
-        self.size_AP = 11  # window size in AP direction (=y) (in voxel)
-        self.size_RL = 1  # window size in RL direction (=x) (in voxel)
-        self.size_IS = 19  # window size in IS direction (=z) (in voxel)
-        self.shift_AP_visu = 15  # shift AP for displaying disc values
-
-    # update constructor with user's parameters
-    def update(self, param_user):
-        for param in param_user.split(','):
-            try:
-                key, value = param.split('=', maxsplit=1)
-                value = int(value)
-            except ValueError:
-                printv('ERROR: Wrong usage.', 1, type='error')
-                sys.exit(1)
-            if key in ['shift_AP', 'size_AP', 'size_RL', 'size_IS', 'shift_AP_visu']:
-                setattr(self, key, value)
-            elif key == 'gaussian_std':
-                # TODO: remove 'gaussian_std' completely for release 6.0
-                printv('WARNING: gaussian_std parameter is currently ignored, '
-                       'and will be removed in a later version.', 1, type='warning')
-            else:
-                printv('ERROR: Wrong usage.', 1, type='error')
-                sys.exit(1)
+# for vertebral_detection
+param_default = {
+    'shift_AP': 32,  # shift the centerline towards the spine (in voxel).
+    'size_AP': 11,  # window size in AP direction (=y) (in voxel)
+    'size_RL': 1,  # window size in RL direction (=x) (in voxel)
+    'size_IS': 19,  # window size in IS direction (=z) (in voxel)
+    'shift_AP_visu': 15,  # shift AP for displaying disc values
+}
+
+
+def vertebral_detection_param(string):
+    """Custom parser for vertebral_detection advanced parameters."""
+    param = param_default.copy()
+    for key_value in string.split(','):
+        key, value = key_value.split('=', maxsplit=1)
+        if key in param:
+            param[key] = int(value)
+        elif key == 'gaussian_std':
+            # TODO(issue#3706): remove 'gaussian_std' completely for v5.7
+            printv('WARNING: gaussian_std parameter is currently ignored, '
+                   'and will be removed in a later version.', 1, type='warning')
+        else:
+            raise ValueError(f'Unknown parameter: {key}')
```

Comment:
I like the refactoring you've done! I noticed, though, that with the newly-proposed method for parsing, the failure given by `argparse` is a little opaque. Some (separate) examples:

```
sct_label_vertebrae: error: argument -param: invalid vertebral_detection_param value: 'test'
sct_label_vertebrae: error: argument -param: invalid vertebral_detection_param value: 'test=None,shift_AP=None'
sct_label_vertebrae: error: argument -param: invalid vertebral_detection_param value: 'shift_AP=None,shift_AP=None'
sct_label_vertebrae: error: argument -param: invalid vertebral_detection_param value: 'shift_AP=9,size_AP=6.0'
```

Regardless of what the actual error is, `argparse` interprets it as a generic type-related error and obscures what the actual issue is.

For niche parsing requirements like this, I tend to like using a [custom action](https://me.micahrl.com/blog/python-argparse-custom-actions-types/)? For example:

```python
import argparse

class ParseParamAction(argparse.Action):
    """Custom parser for vertebral_detection advanced parameters."""
    def __call__(self, parser, namespace, values, option_string=None):
        try:
            param = param_default.copy()
            for key_value in values.split(','):
                key, value = key_value.split('=', maxsplit=1)
                if key in param:
                    param[key] = int(value)
                elif key == 'gaussian_std':
                    # TODO(issue#3706): remove 'gaussian_std' completely for v5.7
                    printv('WARNING: gaussian_std parameter is currently ignored, '
                           'and will be removed in a later version.', 1, type='warning')
                else:
                    raise ValueError(f'Unknown parameter for "-param" argument: {key}')
            return param
        except Exception as e:
            parser.error(e)

    optional.add_argument(
        '-param',
        metavar=Metavar.list,
        action=ParseParamAction,
        default=param_default.copy(),
        help=f'Advanced parameters. Assign value with "="; Separate arguments with ","\n'
             f'  - shift_AP [mm]: AP shift of centerline for disc search (default: {param_default["shift_AP"]})\n'
             f'  - size_AP [mm]: AP window size for disc search (default: {param_default["size_AP"]})\n'
             f'  - size_RL [mm]: RL window size for disc search (default: {param_default["size_RL"]})\n'
             f'  - size_IS [mm]: IS window size for disc search (default: {param_default["size_IS"]})\n',
    )
```

With this, `argparse` will raise a more specific error while still showing the `-h`:

![Screenshot from 2022-03-04 20-47-33](https://user-images.githubusercontent.com/16181459/156862775-bf355e0f-5b9f-4d18-adce-394c5681bb8d.png)

Reply:
Admittedly, though, custom actions aren't well documented in `argparse`'s documentation, though. So, my apologies for throwing a (slightly) arcane pattern at you, aha. And, feel free to tweak this suggestion if needed!

Reply:
That's a very fair point about the specificity of error messages, thanks for bringing it up.

Indeed, my previous draft used a custom Action, yes! I abandoned it because:
1. I didn't know about `parser.error`, so the errors raised by my custom Action had an ugly traceback and no usage message. Now I know!
2. I was uneasy about the handling of the other arguments to `Action.__init__`, like `nargs`, that don't really make sense for this custom Action.

But, good news, I just found that it's possible to have custom error messages from a type function! We just have to use `argparse.ArgumentTypeError` (not to be confused with `argparse.ArgumentError`):
```py3
>>> import argparse
>>> p = argparse.ArgumentParser('my_program')
>>> def f(string):
...     raise argparse.ArgumentTypeError(f'a custom message, {len(string)}')
... 
>>> _ = p.add_argument('--foo', type=f)
>>> p.parse_args(['--foo', 'lalala'])
usage: my_program [-h] [--foo FOO]
my_program: error: argument --foo: a custom message, 6
```

Given this, I'll keep the simple type function approach, but make better error messages.

Reply:
Perfect!!